File: apina-core/src/test/java/fi/evident/apina/java/reader/ClassReaderUtils.java
package fi.evident.apina.java.reader;

import fi.evident.apina.java.model.JavaClass;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;

public final class ClassReaderUtils {

    private ClassReaderUtils() { }

    public static JavaClass loadClass(Class<?> cl) {
        try (InputStream in = openInputStreamForClassBytes(cl)) {
            return ClassMetadataReader.loadMetadata(in);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private static InputStream openInputStreamForClassBytes(Class<?> cl) throws IOException {
        String resourceName = cl.getName().replace('.', '/') + ".class";
        InputStream inputStream = cl.getClassLoader().getResourceAsStream(resourceName);
        if (inputStream != null)
            return inputStream;
        else
            throw new FileNotFoundException(resourceName);
    }
}

File: apina-core/src/test/kotlin/fi/evident/apina/java/reader/ClassReaderUtils.kt
package fi.evident.apina.java.reader

import fi.evident.apina.java.model.JavaClass
import java.io.FileNotFoundException
import java.io.InputStream

fun loadClass(cl: Class<*>): JavaClass =
    openInputStreamForClassBytes(cl).use { ClassMetadataReader.loadMetadata(it) }

private fun openInputStreamForClassBytes(cl: Class<*>): InputStream {
    val resourceName = cl.name.replace('.', '/') + ".class"
    val classLoader = cl.classLoader ?: ClassLoader.getSystemClassLoader()

    return classLoader.getResourceAsStream(resourceName) ?: throw FileNotFoundException(resourceName)
}
